vertx-web/src/main/java/io/vertx/ext/web/impl/RoutingContextImplBase.java

```diff
@@ -106,31 +106,38 @@ protected boolean iterateNext() {
       RouteImpl route = iter.next();
       currentRouteNextHandlerIndex.set(0);
       currentRouteNextFailureHandlerIndex.set(0);
-      if (route.matches(this, mountPoint(), failed)) {
-        if (log.isTraceEnabled()) log.trace("Route matches: " + route);
-        try {
-          currentRoute = route;
-          if (log.isTraceEnabled()) log.trace("Calling the " + (failed ? "failure" : "") + " handler");
-          if (failed && currentRoute.hasNextFailureHandler(this)) {
-            currentRouteNextFailureHandlerIndex.incrementAndGet();
-            route.handleFailure(this);
-          } else if (currentRoute.hasNextContextHandler(this)) {
-            currentRouteNextHandlerIndex.incrementAndGet();
-            route.handleContext(this);
-          } else {
-            continue;
-          }
-        } catch (Throwable t) {
-          if (log.isTraceEnabled()) log.trace("Throwable thrown from handler", t);
-          if (!failed) {
-            if (log.isTraceEnabled()) log.trace("Failing the routing");
-            fail(t);
-          } else {
-            // Failure in handling failure!
-            if (log.isTraceEnabled()) log.trace("Failure in handling failure");
-            unhandledFailure(-1, t, route.router());
+      try {
+        if (route.matches(this, mountPoint(), failed)) {
+          if (log.isTraceEnabled()) log.trace("Route matches: " + route);
+          try {
+            currentRoute = route;
+            if (log.isTraceEnabled()) log.trace("Calling the " + (failed ? "failure" : "") + " handler");
+            if (failed && currentRoute.hasNextFailureHandler(this)) {
+              currentRouteNextFailureHandlerIndex.incrementAndGet();
+              route.handleFailure(this);
+            } else if (currentRoute.hasNextContextHandler(this)) {
+              currentRouteNextHandlerIndex.incrementAndGet();
+              route.handleContext(this);
+            } else {
+              continue;
+            }
+          } catch (Throwable t) {
+            if (log.isTraceEnabled()) log.trace("Throwable thrown from handler", t);
+            if (!failed) {
+              if (log.isTraceEnabled()) log.trace("Failing the routing");
+              fail(t);
+            } else {
+              // Failure in handling failure!
+              if (log.isTraceEnabled()) log.trace("Failure in handling failure");
+              unhandledFailure(-1, t, route.router());
+            }
           }
+          return true;
         }
+      } catch (Throwable t)  {
```

Comment:
Potentially, instead of catching exception in RouteImpl, it can happen in here and then `fail(400)` invoked. Possibly with trace logging, Throwable as a last resort ...

Reply:
The fact is that we can't invoke `context.fail()` because it restarts the routing process, calling again `ctx.next()` and `ctx.matches()` re-throwing exception again.

Reply:
In fact this case is similar to this one: https://github.com/vert-x3/vertx-web/pull/801

Reply:
got it.